File: src/familiar_connect/processors/history_writer.py
"""Single-writer history persistence processor.

Consumes ``discord.text`` events and writes user turns to the
:class:`HistoryStore`. Centralising writes here keeps the SQLite
connection owned by one task and makes dedup-by-``event_id`` simple.

See plan § Design.5 *Side-indices off watermarks* and plan § Design.6
*SQLite write path under load*.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from familiar_connect import log_style as ls
from familiar_connect.bus.topics import TOPIC_DISCORD_TEXT

if TYPE_CHECKING:
    from familiar_connect.bus.envelope import Event
    from familiar_connect.bus.protocols import EventBus
    from familiar_connect.history.store import HistoryStore
    from familiar_connect.identity import Author

_logger = logging.getLogger("familiar_connect.processors.history_writer")


class HistoryWriter:
    """Persist turn-generating events into :class:`HistoryStore`."""

    name: str = "history-writer"
    topics: tuple[str, ...] = (TOPIC_DISCORD_TEXT,)

    def __init__(self, *, store: HistoryStore, familiar_id: str) -> None:
        self._store = store
        self._familiar_id = familiar_id
        # in-process dedup set; survives a single run. Acceptable because
        # the bus itself doesn't republish on retry today.
        self._seen: set[str] = set()

    async def handle(self, event: Event, bus: EventBus) -> None:  # noqa: ARG002
        if event.topic != TOPIC_DISCORD_TEXT:
            return
        if event.event_id in self._seen:
            return

        payload = event.payload
        if not isinstance(payload, dict):
            return
        if payload.get("familiar_id") != self._familiar_id:
            return

        channel_id = payload.get("channel_id")
        content = payload.get("content") or ""
        if not isinstance(channel_id, int) or not content:
            return

        author: Author | None = payload.get("author")
        guild_id = payload.get("guild_id")

        self._seen.add(event.event_id)
        self._store.append_turn(
            familiar_id=self._familiar_id,
            channel_id=channel_id,
            role="user",
            content=content,
            author=author,
            guild_id=guild_id if isinstance(guild_id, int) else None,
        )
        _logger.debug(
            f"{ls.tag('History', ls.LC)} "
            f"{ls.kv('append', event.event_id, vc=ls.LC)} "
            f"{ls.kv('channel', str(channel_id), vc=ls.LY)} "
            f"{ls.kv('text', ls.trunc(content, 80), vc=ls.LW)}"
        )

File: familiar-connect/src/processors/history_writer.rs
//! Single-writer history persistence processor (subsystem 06; Python
//! `processors/history_writer.py`).
//!
//! Consumes `discord.text` events and writes user turns to the store. This is
//! the legacy standalone writer — **not** wired alongside `TextResponder` in
//! production (which owns the user-turn write for read-after-write
//! consistency), but retained, exported, and fully tested.

use std::collections::HashSet;
use std::sync::{Arc, Mutex};

use crate::bus::envelope::Event;
use crate::bus::protocols::EventBus;
use crate::bus::topics::TOPIC_DISCORD_TEXT;
use crate::history::async_store::AsyncHistoryStore;
use crate::history::store::AppendTurn;
use crate::log_style as ls;
use crate::processors::DiscordTextPayload;

/// Persists turn-generating events into the history store.
pub struct HistoryWriter {
    store: Arc<AsyncHistoryStore>,
    familiar_id: String,
    /// In-process dedup set; survives a single run (the bus does not republish).
    seen: Mutex<HashSet<String>>,
}

impl HistoryWriter {
    /// The processor's human label.
    pub const NAME: &'static str = "history-writer";

    /// Construct a writer for `familiar_id` persisting into `store`.
    #[must_use]
    pub fn new(store: Arc<AsyncHistoryStore>, familiar_id: impl Into<String>) -> Self {
        Self {
            store,
            familiar_id: familiar_id.into(),
            seen: Mutex::new(HashSet::new()),
        }
    }

    /// The processor name (`"history-writer"`).
    #[must_use]
    pub const fn name(&self) -> &'static str {
        Self::NAME
    }

    /// The subscribed topics (`discord.text`).
    #[must_use]
    pub const fn topics(&self) -> [&'static str; 1] {
        [TOPIC_DISCORD_TEXT]
    }

    /// Handle one event: dedup + drop rules, then append a user turn.
    ///
    /// # Errors
    /// Propagates a store write failure.
    pub async fn handle(&self, event: &Event, _bus: &dyn EventBus) -> anyhow::Result<()> {
        if event.topic != TOPIC_DISCORD_TEXT {
            return Ok(());
        }
        if self
            .seen
            .lock()
            .expect("history writer seen mutex")
            .contains(&event.event_id)
        {
            return Ok(());
        }
        let Some(payload) = event.payload.downcast_ref::<DiscordTextPayload>() else {
            return Ok(());
        };
        if payload.familiar_id != self.familiar_id {
            return Ok(());
        }
        if payload.content.is_empty() {
            return Ok(());
        }

        self.seen
            .lock()
            .expect("history writer seen mutex")
            .insert(event.event_id.clone());

        let mut append = AppendTurn::new(
            &self.familiar_id,
            payload.channel_id,
            "user",
            &payload.content,
        );
        append = append.pings_bot(payload.pings_bot);
        if let Some(author) = &payload.author {
            append = append.author(author.clone());
        }
        if let Some(guild_id) = payload.guild_id {
            append = append.guild_id(guild_id);
        }
        self.store.append_turn(append).await?;
        tracing::debug!(
            "{} {} {} {}",
            ls::tag("History", ls::LC),
            ls::kv_styled("append", &event.event_id, ls::W, ls::LC),
            ls::kv_styled("channel", &payload.channel_id.to_string(), ls::W, ls::LY),
            ls::kv_styled("text", &ls::trunc(&payload.content, 80), ls::W, ls::LW),
        );
        Ok(())
    }
}
